bokeh/state.py

```diff
@@ -0,0 +1,490 @@
+            specifying which browser to display in, e.g. "safari", "firefox",
+            "opera", "windows-default" (see the ``webbrowser`` module
+            documentation in the standard lib for more details).
+
+        new (str, optional) : new file output mode (default: "tab")
+            For file-based output, opens or raises the browser window
+            showing the current output file.  If **new** is 'tab', then
+            opens a new tab. If **new** is 'window', then opens a new window.
+
+    .. note::
+        The ``browser`` and ``new`` parameters are ignored when showing in
+        an IPython/Jupyter notebook.
+
+    """
+    _show_with_state(obj, _state, browser, new)
+
+_new_param = {'tab': 2, 'window': 1}
+
+def _show_with_state(obj, state, browser, new):
+
+    controller = browserlib.get_browser_controller(browser=browser)
+
+    if state.notebook:
+        _show_notebook_with_state(obj, state)
+
+    elif state.session:
+        _show_server_with_state(obj, state, new, controller)
+
+    if state.file:
+        _show_file_with_state(obj, state, new, controller)
+
+def _show_file_with_state(obj, state, new, controller):
+    save(obj, state=state)
+    controller.open("file://" + os.path.abspath(state.file['filename']), new=_new_param[new])
+
+def _show_notebook_with_state(obj, state):
+    if state.session:
+        push(state=state)
+        snippet = autoload_server(obj, state.session)
+        publish_display_data({'text/html': snippet})
+    else:
+        publish_display_data({'text/html': notebook_div(obj)})
+
+def _show_server_with_state(obj, state, new, controller):
+    push(state=state)
+    controller.open(state.session.object_link(state.document.context), new=_new_param[new])
+
+def save(obj, filename=None, resources=None, title=None):
+    """ Save an HTML file with the data for the current document.
+
+    Will fall back to the default output state (or an explicitly provided
+    ``State`` object) for ``filename``, ``resources``, or ``title`` if they
+    are not provided.
+
+    Args:
+        obj (Document or Widget/Plot object)
+
+        filename (str, optional) : filename to save document under
+
+        resources (Resources, optional) : BokehJS resource config to use
+
+        title (str, optional) : title of the bokeh plot
+
+    Returns:
+        None
+
+    """
+    filename, resources, title = _get_save_args(_state, filename, resources, title)
+
+    _save_helper(obj, filename, resources, title)
+
+def _get_save_args(state, filename, resources, title):
+
+    if filename is None and state.file:
+        filename = state.file['filename']
+
+    if resources is None and state.file:
+        resources = state.file['resources']
+
+    if title is None and state.file:
+        title = state.file['title']
+
+    if not filename:
+        warnings.warn("save() called but no filename was supplied and output_file(...) was never called, nothing saved")
```

Comment:
Just wondering, since I saw this elsewhere. When should we use `warning.warn()` and when `logging.warn`? What I saw is that `warning.warn` does not pass through `bokeh.logging`.